File: src/#160/solution.go
package solution

type ListNode struct {
	Val  int
	Next *ListNode
}

func getIntersectionNode(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}
	cnt := 0
	currentA := headA
	currentB := headB
	for currentA != nil || currentB != nil {
		if currentA != nil {
			cnt++
			currentA = currentA.Next
		}
		if currentB != nil {
			cnt--
			currentB = currentB.Next
		}
	}
	if cnt > 0 {
		currentA, currentB = headA, headB
	} else {
		currentA, currentB, cnt = headB, headA, -cnt
	}
	for i := cnt; i > 0; i-- {
		currentA = currentA.Next
	}
	for currentA != currentB {
		currentA = currentA.Next
		currentB = currentB.Next
	}
	return currentA
}

File: src/#160/solution.js
function ListNode(val) {
  this.val = val
  this.next = null
}

var getIntersectionNode = function (headA, headB) {
  if (headA === null || headB === null) {
    return null
  }

  let cnt = 0
  let currentA = headA
  let currentB = headB
  
  while (currentA !== null || currentB !== null) {
    if (currentA !== null) {
      cnt++
      currentA = currentA.next
    }
    if (currentB !== null) {
      cnt--
      currentB = currentB.next
    }
  }

  if (cnt > 0) {
    currentA = headA
    currentB = headB
  } else {
    currentA = headB
    currentB = headA
    cnt = -cnt
  }

  for (let i = cnt; i > 0; i--) {
    currentA = currentA.next
  }

  while (currentA !== currentB) {
    currentA = currentA.next
    currentB = currentB.next
  }
  return currentA
}
